File: LeetCode/207. Course Schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prereqs) {
        Map<Integer, List<Integer>> preMap = new HashMap<>();
        Set<Integer> visit = new HashSet<>();

        class H {
            boolean dfs(int c) {
                if (preMap.isEmpty()) return true;
                if (visit.contains(c)) return false;

                visit.add(c);
                if (preMap.containsKey(c)) {
                    for (int pre : preMap.get(c)) {
                        if (!dfs(pre)) return false;
                    }
                }
                visit.remove(c);
                if (preMap.get(c) == null) {
                    preMap.put(c, new ArrayList<>());
                } else preMap.get(c).clear();
                return true;
            }
        }

        for (int[] pre : prereqs) {
            List<Integer> list = preMap.get(pre[0]);
            if (list == null) {
                preMap.put(pre[0], new ArrayList<>());
                list = preMap.get(pre[0]);
            }
            list.add(pre[1]);
        }
        for (int c=0;c<numCourses;c++) {
            if (!new H().dfs(c)) return false;
        }
        return true;
    }
} //TC: O(n+e), SC: O(n+e), where n is numCourses and e is the size of prereqs

File: LeetCode/207. Course Schedule.kt
class Solution {
    val preMap = HashMap<Int, MutableList<Int>>() //course:prereq list
    val visitSet = HashSet<Int>() //dfs visited courses

    fun canFinish(numCourses: Int, prereqs: Array<IntArray>): Boolean {
        prereqs.forEach { //map prerequisites
            if (it[0] in preMap) { //add new prerequisites if the course is already in the map
                preMap[it[0]] = preMap[it[0]]!!.apply{add(it[1])}
            } else { //add new prerequisite list if it isn't
                preMap[it[0]] = mutableListOf(it[1])
            }
        }
        for (c in 0 until numCourses) { //loop all courses
            if (!dfs(c)) return false
        }
        return true
    }

    fun dfs(course: Int): Boolean {
        if (course in visitSet) return false //if the course is already visited, that means there is a loop, so returning false
        if (preMap.isEmpty()) return true //if there is no prerequisite, return true

        visitSet.add(course) //add the course to visit set
        if (course in preMap) {
            for(p in preMap[course]!!) { //loop prerequisites of the course
                if (!dfs(p)) return false //return false if there is a loop
            }
        }
        visitSet.remove(course) //remove the course if no loop is found
        preMap[course] = mutableListOf<Int>() //reset the prerequisite map for the course
        return true
    }
}
